Allow cache_belongs_to on polymorphic associations.

dylanahsmith opened the pull request:
@alexandcote does this fix #225 for you?
@lolownia & @lrajlich does this fix #215 for you?

## Problem

In cache_belongs_to we were assuming that a belongs_to association has a single association class that could be retrieved from the association reflection, but this isn't the case for polymorphic belongs_to associations, where a record is needed to get the actual class.

## Solution

Get the class from the association rather than the association reflection in the `fetch_#{association}` method.

This still needs some regression tests.

dylanahsmith commented on lib/identity_cache/query_api.rb:
```diff
@@ -259,11 +259,16 @@ def prefetch_associations(associations, records)
             else
+              reflection = details[:association_reflection]
+              if reflection.options[:polymorphic]
+                raise ArgumentError.new("Polymorphic belongs_to associations do not support prefetching yet.")
```
This is for simplicity, since when dealing with multiple records, we might have a mix of types to fetch.  It could be implemented by grouping by type to fetch them, but I'm not sure if we want to support that complexity at the moment.

alexandcote commented:
Yeah ! This work perfectly for me ! Thanks a lot @dylanahsmith for this fix

lolownia commented:
Unfortunately no, sorry. Did not fix #215 
Please check out a minimal example here:
https://github.com/lolownia/identity_cache_bug
so far I can workaround this bug by setting autoload in config/initializers like:
```
autoload :Foo, File.expand_path('../../../models/foo.rb', __FILE__)
```

dylanahsmith commented:
> so far I can workaround this bug by setting autoload in config/initializers

issue #215 is about a circular references, not about expecting identity_cache to implement autoloading for you, so it sounds like this would fix that issue.

dylanahsmith commented:
Updated the documentation to include the example usage of a polymorphic cache_belongs_to

I added a regression test for issue #225.  I'm not sure what is a good way to test issue #215, so there is no regression test for that.

dylanahsmith commented:
@byroot & @fbogsany would either of you mind giving this a review?

dylanahsmith commented:
> I'm not sure what is a good way to test issue #215, so there is no regression test for that.

I guess a test explicitly for this isn't needed, since calling `reflect_on_association(association).klass` in cache_belongs_to would fail on a polymorphic association, as is tested in test_fetching_polymorphic_belongs_to_association

byroot commented:
I can't right now but îm flaging et t

byroot commented:
:clap: LGTM

rafaelfranca commented:
:shipit:
